lib/internal/fs/promises.js

```diff
@@ -663,19 +666,55 @@ async function writeFile(path, data, options) {
   options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'w' });
   const flag = options.flag || 'w';
 
-  if (!isArrayBufferView(data)) {
-    validateStringAfterArrayBufferView(data, 'data');
-    data = Buffer.from(data, options.encoding || 'utf8');
-  }
+  if (isIterable(data)) {
+    if (options.signal?.aborted) {
+      throw lazyDOMException('The operation was aborted', 'AbortError');
+    }
+    const fd = await open(path, flag, options.mode);
+    try {
+      if (options.signal?.aborted) {
+        throw lazyDOMException('The operation was aborted', 'AbortError');
+      }
+      for await (const buf of data) {
+        if (options.signal?.aborted) {
+          throw lazyDOMException('The operation was aborted', 'AbortError');
+        }
+        await fd.write(buf);
+        if (options.signal?.aborted) {
+          throw lazyDOMException('The operation was aborted', 'AbortError');
+        }
```

Comment:
IMO there's no need to abort if this finished writing before it saw that it was aborted (so just keep the first check)

Reply:
I suggested to add this check to stop the iteration: if the signal is aborted while writing, we shouldn't wait to the next iteration of `data` to throw.

Reply:
fair enough, it's just kind of weird if the iterator already finished